src/avatar.cpp

```diff
@@ -1142,3 +1143,115 @@ void avatar::vomit()
     }
     Character::vomit();
 }
+
+int avatar::get_str_base() const
+{
+    return Character::get_str_base() + std::max( 0, str_upgrade );
+}
+
+int avatar::get_dex_base() const
+{
+    return Character::get_dex_base() + std::max( 0, dex_upgrade );
+}
+
+int avatar::get_int_base() const
+{
+    return Character::get_int_base() + std::max( 0, int_upgrade );
+}
+
+int avatar::get_per_base() const
+{
+    return Character::get_per_base() + std::max( 0, per_upgrade );
+}
+
+int avatar::kill_xp() const
+{
+    // TODO: game::kills probably should be avatar::kills
+    return g->kill_xp();
+}
+
+// based on  D&D 5e level progression
+static const std::vector<int> xp_cutoffs = {
+    300, 900, 2700, 6500, 14000,
+    23000, 34000, 48000, 64000, 85000,
+    100000, 120000, 140000, 165000, 195000,
+    225000, 265000, 305000, 355000, 405000
+};
+
+int avatar::free_upgrade_points() const
+{
+    const int xp = kill_xp();
+    int lvl = 0;
+    for( const int &xp_lvl : xp_cutoffs ) {
+        if( xp >= xp_lvl ) {
+            lvl++;
+        } else {
+            break;
+        }
+    }
+    return lvl - str_upgrade - dex_upgrade - int_upgrade - per_upgrade;
+}
+
+static int xp_to_next()
+{
+    const int cur_xp = g->u.kill_xp();
+    int xp_next = 0;
+    for( std::vector<int>::const_iterator iter = xp_cutoffs.cbegin(); iter != xp_cutoffs.cend();
+         iter++ ) {
+        if( cur_xp >= *iter ) {
+            xp_next = *iter;
+        } else {
+            if( iter + 1 == xp_cutoffs.end() ) {
+                xp_next = -1;
+            } else {
+                xp_next = *( iter + 1 );
+            }
+            break;
+        }
+    }
+    return xp_next;
+}
+
+void avatar::upgrade_stat_prompt( const Character::stat &stat )
+{
+    const int free_points = free_upgrade_points();
+    const int next_lvl_xp = xp_to_next();
+    std::string stat_string;
+    switch( stat ) {
+        case STRENGTH:
+            stat_string = _( "strength" );
+            break;
+        case DEXTERITY:
+            stat_string = _( "dexterity" );
+            break;
+        case INTELLIGENCE:
+            stat_string = _( "intelligence" );
+            break;
+        case PERCEPTION:
+            stat_string = _( "perception" );
+            break;
+        default:
+            return;
+    }
+    if( free_points == 0 ) {
```

Comment:
You can pull this up and return after `popup`. Then you don't need `else if`.

Reply:
i believe i've addressed what you are talking about